Fixing the Tomcat state's indifference to a lack of versions and incorrect file paths

thor opened the pull request:
### What does this PR do?

This pull request aims to solve and allow for deployed applications to context paths _without_ having a blank version number, and allows you to disable extracting the version out of the filename.
### What issues does this PR fix or reference?

Per the commit it aims to fix #24990.
### Previous Behavior

The main issue previously were that deploying a file named `jenkins.war` would result in a deployment with a _empty_ version, and as such it would be deployed to `jenkins##`. The same behaviour would be present if you deployed `jenkins-1.2.4.war` and you manually specified `version: ''` for in the state `tomcat.war_deployed`.
### New Behavior

The main _backwards affecting_ change is that now WARs without a version in them--or specified--will have a correct file path and be without a version string in Tomcat's eyes. For instance, in this case deploying `jenkins.war` will be deployed to `jenkins` inside the Tomcat webapps folder (e.g. `/usr/lib/tomcat/webapps/` on CentOS 7 at least).

However, this means that if you have `jenkins.war` deployed to `jenkins##` at the moment, **this change will redeploy `jenkins.war` from `jenkins##` to `jenkins`**.

Furthermore, if you're deploying a file like `jenkins-1.2.4.war`, you can now disable extracting the version from the filename by specifying `version: False`. (You _must_ use `version: False`, as the `version` parameter is an empty string by default.)

My question would be, could this be **released in a bugfix update or is it too backwards incompatible?** I'm slightly torn: on one hand it never should've deployed to `blah##` in the first place, and WAR-files usually don't much to configuration in them you do manually (and you'd think that'd be controlled with states anyway), _but_ it could cause someone to go "huh". Thoughts?
### Tests written?

Yes, the current tests were slightly updated and some new ones to deal specifically with _no version_ to _version_ and _version_ to _no version_, as well as _no version_ to _blank version_ and such.

From the commit:

> Ensure that lack of version does not result in blank version being used,
> as to avoid certain Java EE applications from having issues with paths
> when parallell execution is not "supported".
> 
> This fixes saltstack/salt#24990 by both not adding a blank version if there is no
> version to exract, but also by allowing you to overwrite that there
> should be _no_ version whether or not your WAR-file ends with `-1.2.3`.
> 
> Tests added for specifically dealing with no-to-version scenarios,
> as well as several minor documentation tweaks that hopefully aren't too
> interruptive!

thor committed:
Initial work on adapting Tomcat to care correctly

Ensure that lack of version does not result in blank version being used,
as to avoid certain Java EE applications from having issues with paths
when parallell execution is not "supported".

This fixes #24990 by both not adding a blank version if there is no
version to exract, but also by allowing you to overwrite that there
should be *no* version whether or not your WAR-file ends with `-1.2.3`.

Tests added for specifically dealing with no-to-version scenarios,
as well as several minor documentation tweaks that hopefully aren't too
interruptive!

cachedout commented:
While I agree that this is technically a bugfix, I think that the change in behavior should not be present in a point release. I'd like to keep this in the develop branch and if somebody really needs it, they can put the module and state in their own deployment.

cachedout commented:
Before we get this merged, we do need the lint errors fixed up those please. Thanks, @thor https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/3726/violations/file/salt/modules/tomcat.py/

thor committed:
Fixing coding style and change w/o updated test

Fixed testing for tomcat.wait, as well as updating style. Should resolve
any test errors isolated to the module and/or state.

thor commented:
Thanks for the feedback, @cachedout! 👍 I've fixed the linting issues reported, as well as a failing test after I updated a string in the state.

I've personally just lazy'd a bit and used the state and module in our `_states/`/`_modules` folders in our roots with fairly okay results, where only `_extract_war_version` is problematic because it's imported from the Salt package itself.

cachedout commented:
@thor If you'd like to make that change, then I'll put it in a bug-fix branch. Shall I hold off a bit?

thor commented:
@cachedout I wanted to, but it seems it's more hassle than it's worth -- too much dealing with special cases and I feel like a good solution to it would take some time. I'd recommend just merging this into the next release. :) Anyone who wants to use this can--after all--simply put `states/tomcat.py` into `_states/tomcat.py` in their roots.

I'd do some bigger documentation changes, but I'm struggling with failing to build it, so I can just PR that some other time. :) Note the versionadded-tag (or should it ideally be versionchanged?) needs an upcoming version number.

(*Merge ahoy!*)

cachedout commented:
Great! Thanks so much, @thor
